Add windows named pipe to dog statsd

ogaca-dd opened the pull request:
### What does this PR do?

Add a Windows named pipe listener for DogStatsd.
Named pipe uses a connected mode:
  - There is a Go routine that listen for new clients.
  - When a new client connects, a new go routines is run to read the message coming from this go routine. 

This PR also extract some common logics for DogStatsd listeners.
 
**This feature is experimental and not officially supported for now**
### Motivation

Using UDP is not possible on Azure Web Service for DogStatsd.

### Additional Notes

This PR does not impact existing Dogstatsd listeners (UDP and Unix domain socket).


### Describe your test plan

I wrote a C# client that send metric from multiple threads. I also checked the connection is working after a reconnection.

*Name pipes are not supported on Windows 7:* https://github.com/microsoft/go-winio/pull/174

ogaca-dd committed:
Add NamedPipeListener

ogaca-dd committed:
Rename namedPipeTelemetry to listenerTelemetry

ogaca-dd committed:
Make listenerTelemetry generic

ogaca-dd committed:
Use listenerTelemetry inside UDPListener

ogaca-dd committed:
Use listenerTelemetry inside UDSListener

ogaca-dd committed:
Extract packetAssembler logic from udp to listener_package.

ogaca-dd committed:
Use listenerPacket inside NamedPipeListener

ogaca-dd committed:
Refactor NewNamedPipeListener

ogaca-dd committed:
Make listenerTelemetry static.

ogaca-dd committed:
Make NamedPipeListener.connections thread safe.

ogaca-dd committed:
Add unit tests for NamedPipeListener.

ogaca-dd committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into olivierg/AddWindowsNamedPipeToDogStatsd

ogaca-dd commented on pkg/dogstatsd/listeners/udp.go:
```diff
@@ -96,7 +78,6 @@ func NewUDPListener(packetOut chan Packets, sharedPacketPool *PacketPool) (*UDPL
 func (l *UDPListener) Listen() {
 	log.Infof("dogstatsd-udp: starting to listen on %s", l.conn.LocalAddr())
 	for {
-		udpPackets.Add(1)
```
The behavior is not exactly the same as before but IMO counting udp packets should be done after `ReadFrom`.

ogaca-dd commented on pkg/dogstatsd/listeners/uds_common.go:
```diff
@@ -19,28 +19,10 @@ import (
-		nil, "Dogstatsd UDS origin detection error count")
-
-	tlmUDSPacketsBytes = telemetry.NewCounter("dogstatsd", "uds_packets_bytes",
-		nil, "Dogstatsd UDS packets bytes")
```
The previous message was `"Dogstatsd UDS packets bytes"`. Is it a problem?

ogaca-dd committed:
Minor improvements.

ogaca-dd committed:
Rename listenerPacket packetManager

remeh commented on pkg/dogstatsd/listeners/named_pipe_windows.go:
```diff
@@ -0,0 +1,143 @@
+			l.mux.Lock()
+			l.connections[conn] = struct{}{}
+			l.mux.Unlock()
+			go l.listenConnection(conn)
```
Did you have the chance to valide that spawning so many routines isn't a performance issue? I mean, spawning that many routines may not be an issue itself but every routines are competiting for the `l.mux` mutex, and in cases where we receive hundreds of thousands of metrics per seconds, I don't see how this could not be an issue?

ogaca-dd replied:
This code is called when a new client connects to the server. It happens once by client.

remeh replied:
Ah OK I thought it was each time we received a packet, cool!

remeh commented on pkg/dogstatsd/listeners/named_pipe_windows.go:
```diff
@@ -0,0 +1,143 @@
+				log.Infof("dogstatsd-named-pipes: stop listening a named pipe client on %s", conn.LocalAddr())
+				break
+			}
+			log.Errorf("dogstatsd-named-pipe: error reading packet: %v", err.Error())
```
Typo: everywhere else it's `dogstatsd-named-pipes`

ogaca-dd replied:
Fixed.

remeh commented on pkg/dogstatsd/listeners/named_pipe_windows.go:
```diff
@@ -0,0 +1,143 @@
+			l.mux.Unlock()
+			go l.listenConnection(conn)
+
+		case err.Error() == "use of closed network connection":
+			{
+				// Called when the pipe listener is closed from Stop()
+				log.Info("dogstatsd-named-pipes: stop listening")
+				return
+			}
+		default:
+			log.Error(err)
+		}
```
Do you know which cases could lead to the default case? I'm a bit afraid of it spamming the logs with errors.

ogaca-dd replied:
It is a good question, but I do not know the answer.
I use the same error logic as for https://github.com/DataDog/datadog-agent/blob/13edd7fd5d1ea682f630508f4cf60a8974912a15/pkg/dogstatsd/listeners/udp.go#L76.

remeh commented on pkg/dogstatsd/listeners/named_pipe_windows.go:
```diff
@@ -0,0 +1,143 @@
+func (l *NamedPipeListener) listenConnection(conn net.Conn) {
+	log.Infof("dogstatsd-named-pipes: start listening a new named pipe client on %s", conn.LocalAddr())
+	for {
+		n, err := conn.Read(l.packetManager.buffer)
```
Isn't that an issue that all routines share the same buffer here?

ogaca-dd replied:
Yes this is an issue and fixed in the next commit.

ogaca-dd committed:
Use a new buffer for each client in NamedPipeListener.

ogaca-dd commented on pkg/dogstatsd/listeners/uds_common.go:
```diff
@@ -19,28 +19,23 @@ import (
-	tlmUDSOriginDetectionError = telemetry.NewCounter("dogstatsd", "uds_origin_detection_error",
-		nil, "Dogstatsd UDS origin detection error count")
+type udsListenerTelemetry struct {
+	*listenerTelemetry
```
It is because `newListenerTelemetry` returns `*listenerTelemetry` (listenerTelemetry has 6 field).

ogaca-dd commented on pkg/dogstatsd/listeners/uds_common.go:
```diff
@@ -19,28 +19,23 @@ import (
-	tlmUDSPacketsBytes = telemetry.NewCounter("dogstatsd", "uds_packets_bytes",
-		nil, "Dogstatsd UDS packets bytes")
-)
+var udsTelemetry *udsListenerTelemetry
```
Fixed.

ogaca-dd committed:
Handle review comments.

ogaca-dd commented on pkg/dogstatsd/listeners/packet_manager.go:
```diff
@@ -0,0 +1,47 @@
+	"github.com/DataDog/datadog-agent/pkg/config"
+)
+
+type packetManager struct {
```
Fixed.

ogaca-dd commented on pkg/dogstatsd/listeners/uds_common.go:
```diff
@@ -133,7 +128,6 @@ func (l *UDSListener) Listen() {
 		// retrieve an available packet from the packet pool,
 		// which will be pushed back by the server when processed.
 		packet := l.sharedPacketPool.Get()
-		udsPackets.Add(1)
```
I added `t.packets.Add(1)` to `onReadError()`. Now, the only difference is that we do not count the packet when the connection is closed. Do you think it is OK?

ogaca-dd committed:
Minor fixed.

ogaca-dd committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into olivierg/AddWindowsNamedPipeToDogStatsd

ogaca-dd committed:
Add config.BindEnvAndSetDefault for `dogstatsd_windows_pipe_name`

ogaca-dd committed:
Revert changes in uds_common.go and udp.go

remeh reviewed (approved):
The CI seems to have issue with the unit tests, except from that 👍

ogaca-dd committed:
Make `packetManager.go` and `listenerTelemetry.go` windows only

ogaca-dd committed:
Add a comment in config.go for dogstatsd_windows_pipe_name.

ogaca-dd committed:
Handle partial messages.

ogaca-dd committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into olivierg/AddWindowsNamedPipeToDogStatsd

ogaca-dd committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into olivierg/AddWindowsNamedPipeToDogStatsd

truthbk commented on pkg/dogstatsd/server.go:
```diff
@@ -163,6 +163,16 @@ func NewServer(aggregator *aggregator.BufferedAggregator) (*Server, error) {
+	if len(pipeName) > 0 {
+		namedPipeListener, err := listeners.NewNamedPipeListener(pipeName, packetsChannel, sharedPacketPool)
+		if err != nil {
+			log.Errorf(err.Error())
```
I think when we error out we'll get an underlying lib error from `winio` (for example) but it might be immediately obvious in the logs that it's from the windows named pipe, I would maybe add something like:
```
log.Errorf("named pipe error: %v", err.Error())
```

truthbk reviewed (commented):
@remeh already did a pretty thorough review, this looks good to me, just a small comment.

ogaca-dd committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into olivierg/AddWindowsNamedPipeToDogStatsd

ogaca-dd committed:
Improve error message.

ogaca-dd committed:
Close all connections before NamedPipeListener.Stop returns.

If a previous connection is opened, pipe.Accept() can raise the following error

panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x18 pc=0xaa36e9]

Note: pipe is not null in this case.

ogaca-dd committed:
Fix a race condition in unit tests.

olivielpeau reviewed (approved):
A few nits, LGTM once addressed

ogaca-dd committed:
Fix TestNamedPipeStop

ogaca-dd commented on pkg/dogstatsd/listeners/named_pipe_windows.go:
```diff
@@ -63,22 +62,64 @@ func newNamedPipeListener(
+			atomic.AddInt32(&l.activeConnCount, 1)
+		case conn := <-l.connToClose:
+			conn.Close()
+			delete(connections, conn)
```
Unit test pass because 
```
for listener.getActiveConnectionsCount() > 1 {	
		time.Sleep(100 * time.Millisecond)	
	}
```
should be 
```
for listener.getActiveConnectionsCount() > 0 {	
		time.Sleep(100 * time.Millisecond)	
	}
```

ogaca-dd committed:
Rename handleConnection to handleConnections

ogaca-dd committed:
Remove unneeded line.

ogaca-dd commented on pkg/process/config/config.go:
```diff
@@ -408,6 +408,7 @@ func loadEnvVariables() {
 		{"DD_ORCHESTRATOR_URL", "process_config.orchestrator_dd_url"},
 		{"DD_HOSTNAME", "hostname"},
 		{"DD_DOGSTATSD_PORT", "dogstatsd_port"},
+		{"DD_DOGSTATSD_WINDOWS_PIPE_NAME", "dogstatsd_windows_pipe_name"},
```
After I added this line, I could set the pipe name with the env var `DD_DOGSTATSD_WINDOWS_PIPE_NAME` and so I believed this mapping was needed.

As `DD_DOGSTATSD_WINDOWS_PIPE_NAME` follow the env var convention, it works thanks for this line https://github.com/DataDog/datadog-agent/pull/5809/files#diff-091ace02b49a46f55f2586cfec89a160R313.

olivielpeau reviewed (approved):
🚀
